dcpy/connectors/socrata/publish.py

```diff
@@ -0,0 +1,350 @@
+"""DCP Socrata Connector to bridge the gaps between various Socrata APIs/SDKs.
+
+The publishing API documentation is here:
+https://dev.socrata.com/docs/other/publishing.html#?route=overview
+
+Additionally, when it's mysterious about which endpoint to hit, or what the
+arguments should be, opening the web interface and watching the requests is
+quite informative.
+"""
+
+from __future__ import annotations
+from dataclasses import dataclass
+import json
+import os
+from pathlib import Path
+import requests
+from socrata.authorization import Authorization
+from socrata import Socrata as SocrataPy
+import typer
+from typing import TypedDict, Literal
+
+from .metadata import datasets
+
+from dcpy.utils.logging import logger
+
+from dcpy.utils import s3
+
+SOCRATA_USER = os.getenv("SOCRATA_USER")
+SOCRATA_PASSWORD = os.getenv("SOCRATA_PASSWORD")
+
+SOCRATA_DOMAIN = "data.cityofnewyork.us"
+SOCRATA_API_EP = f"https://{SOCRATA_DOMAIN}/api"
+_revisions_root = f"{SOCRATA_API_EP}/publishing/v1/revision"
+_views_root = f"{SOCRATA_API_EP}/views"
+
+DISTRIBUTIONS_BUCKET = "edm-distributions"
+
+
+def _simple_auth():
+    return (SOCRATA_USER, SOCRATA_PASSWORD)
+
+
+def _socrata_request(
+    url,
+    method: Literal["GET", "PUT", "POST", "PATCH", "DELETE"],
+    **kwargs,
+) -> dict:
+    """Request wrapper to add auth, and raise exceptions on error."""
+    request_fn = getattr(requests, method.lower())
+    resp = request_fn(url, auth=_simple_auth(), **kwargs)
+    if not resp.ok:
+        raise Exception(resp.text)
+    return resp.json()
+
+
+class Socrata:
+    """Helper classes for working with responses from/inputs to the Socrata Rest APIs.
+
+    Some classes, e.g. the TypedDict subclasses, are here more as documentation.
+    """
+
+    class Inputs:
+        """Helper classes to model inputs to the Socrata API."""
+
+        class Attachment(TypedDict):
+            name: str
+            filename: str
+            asset_id: str
+            blob_id: str | None
+
+    class Responses:
+        """Helper classes to model responses from the Socrata API."""
+
+        class Revision:
+            def __init__(self, data):
+                self._data = data
+                self._raw_resource_data = data["resource"]
+                self.revision_num = self._raw_resource_data["revision_seq"]
+                self.four_four = self._raw_resource_data["fourfour"]
+                self.metadata = self._raw_resource_data["metadata"]
+                self.attachments = self._raw_resource_data["attachments"]
+
+        class RevisionDataSource:
+            def __init__(self, socrata_source):
+                self.socrata_source = socrata_source
+
+                # AR Note: It's not clear to me why/when you would have multiple schemas/output
```

Comment:
Some views on opendata are multiple data sources, could it be a case like that?

https://data.cityofnewyork.us/Housing-Development/Building-Footprints/nqwf-w8eh

One "four_four" (great name for the field) corresponds to two datasets in this case

Reply:
Or is this truly for one data source here, not for one "four_four"

Reply:
Yeah, it's the latter. A Source is an uploaded dataset (say, an individual shapefile) of which you may have multiple... but I don't quite get why there would be multiple schemas for that one thing.